academic_observatory_workflows/workflows/openalex_telescope.py

```diff
@@ -21,590 +21,923 @@
+        self.add_setup_task(self.fetch_releases)
+        self.add_task(self.create_datasets)
+
+        # Create snapshots of main tables in case we mess up
+        # This is done before updating the tables to make sure that the snapshots haven't expired before the tables
+        # are updated
+        self.add_task(self.bq_create_main_table_snapshots)
+
+        # Transfer, download and transform data
+        self.add_task(self.aws_to_gcs_transfer)
+
+        # Download concepts, institutions and works which need to be pre-processed
+        # Gsutil is used instead of the standard Google Cloud Python library, because it is faster at downloading files
+        # than the Google Cloud Python library.
+        for entity_name, transform in self.entities:
+            if transform:
+                output_folder = "{{ release.download_folder }}/data/" + entity_name + "/"
+                bucket_path = "{{ release.gcs_openalex_data_uri }}data/" + entity_name + "/*"
+                self.add_operator(
+                    WorkflowBashOperator(
+                        workflow=self,
+                        task_id=f"download_{entity_name}",
+                        bash_command="mkdir -p "
+                        + output_folder
+                        + " && gcloud auth activate-service-account --key-file=${GOOGLE_APPLICATION_CREDENTIALS}"
+                        + " && gsutil -m -q cp -L {{ release.log_path }} -r "
+                        + bucket_path
+                        + " "
+                        + output_folder,
+                    )
+                )
         self.add_task(self.transform)
-        self.add_task(self.upload_transformed)
-        self.add_task(self.bq_append_new)
-        self.add_task(self.bq_delete_old)
-        self.add_task(self.bq_create_snapshot)
+
+        # Upsert records
+        self.add_task(self.upload_upsert_files)
+        self.add_task(self.bq_load_upsert_tables)
+        self.add_task(self.bq_upsert_records)
+
+        # Delete records
+        self.add_task(self.bq_load_delete_tables)
+        self.add_task(self.bq_delete_records)
+
+        # Add release info to API and cleanup
         self.add_task(self.add_new_dataset_releases)
+        self.add_task(self.cleanup)
 
-    def add_new_dataset_releases(self, release: OpenAlexRelease, **kwargs):
-        """Task to add new DatasetRelease records in the API after the workflow is done.
-        Also cleans up the local files. This has to be done in the same task, because one of the local files is
-        necessary to create the new dataset releases.
+        # The last task that the next DAG run's ExternalTaskSensor waits for.
+        self.add_operator(
+            DummyOperator(
+                task_id=external_task_id,
+            )
+        )
 
-        :param release: Release passed to workflow tasks.
-        :param kwargs: kwargs passed to PythonOperator functions.
+    @property
+    def aws_key(self) -> Tuple[str, str]:
```

Comment:
I like this solution for streamlining airflow connections! Perhaps we should incorporate something similar with other telescopes in the future?

Reply:
Thanks. Would you suggest consolidating these functions in the future?

get_airflow_connection_login
get_airflow_connection_password
get_airflow_connection_url

Reply:
Sure, do you think it'd be better to combine them and choose the output accordingly for each use?

Reply:
Probably. What those functions do is assert that the right parts of the connection were specified, so there might need to be some keyword parameters to turn the checks for specific parameters on and off.